### ISSUE TITLE
Minor data difference

### ISSUE THREAD

**Issue by Turtle-Duck** [NVIDIA/DALI#1012]
Why there are minor different between the data decoded by NVVL in GPU and the data decoded directly by FFMpeg in CPU?

I tested several videos, the data difference in less than about 3 but exist:

from the NVVL:
![222](https://user-images.githubusercontent.com/14868004/60077722-32b53580-975d-11e9-9113-a5ecb073e7b7.png)

from the CPU python image-FFMpeg directly:
![111](https://user-images.githubusercontent.com/14868004/60077751-3f398e00-975d-11e9-9e11-64908907a9b6.png)

I haven't used resize() function, use the same version of FFMpeg, so which parts may leading to this different?

**Comment by JanuszL** [comment]
Hi,
Could you provide exact options you are using for FFmpeg decoding?
Have you tried FFmpeg with CUDA enabled - as it can provide GPU acceleration for the decoding so we know if this is GPU vs CPU implementation, or something else.

**Comment by Turtle-Duck** [comment]
Hi, thanks for your response!

I use this python3.4.0 code:

```
import imageio
imageio.plugins.ffmpeg.download('/usr/local/share/ffmpeg/bin_test/') # the certain version ffmpeg
vid = imageio.get_reader(path, 'ffmpeg')
frame = vid.get_data(index)[:, :, [2, 1, 0]]
```

with imageio (2.4.1), the FFmpeg is compiled by this options:
```
ffmpeg version N-93855-g21832b9 Copyright (c) 2000-2019 the FFmpeg developers
built with gcc 5.2.0 (GCC)
configuration: --prefix=/usr/local/share/ffmpeg_qlh/build --pkg-config-flags=--static --extra-cflags=-I/usr/local/share/ffmpeg_qlh/build/include --extra-ldflags=-L/usr/local/share/ffmpeg_qlh/build/lib --extra-libs=-lpthread --extra-libs=-lm --bindir=/usr/local/share/ffmpeg_qlh/bin --enable-gpl --enable-libfdk-aac --enable-libmp3lame --enable-libopus --enable-libvorbis --enable-libvpx --enable-libx264 --enable-libx265 --enable-nonfree --enable-shared --enable-pic
libavutil 56. 27.100 / 56. 27.100
libavcodec 58. 52.101 / 58. 52.101
libavformat 58. 27.103 / 58. 27.103
libavdevice 58. 7.100 / 58. 7.100
libavfilter 7. 51.100 / 7. 51.100
libswscale 5. 4.101 / 5. 4.101
libswresample 3. 4.100 / 3. 4.100
libpostproc 55. 4.100 / 55. 4.100
Hyper fast Audio and Video encoder
```

I put the ffmpeg to path /usr/local/share/ffmpeg/bin_test/ffmpeg so that same version ffmpeg as compiled in NVVL are used. Maybe this information is useful~

I have not tried FFmpeg with CUDA enabled, I' ll try it and report the result after a while, thanks again for your response!

**Comment by JanuszL** [comment]
NVVL/DALI doesn't use FFmpeg for decoding, only for the stream parsing so it doesn't matter that much. I can only guess it could be how actual decoding is implemented in CPU vs GPU or FFmpeg is applying some additional post processing or using different decoding options.
If you could provide some self-contained reproduction script that we can use to get your results (for example using https://github.com/NVIDIA/DALI_extra/tree/master/db/optical_flow/sintel_trailer)?

**Comment by Turtle-Duck** [comment]
Thanks!

I upload some scripts and the test video in [here](https://github.com/Turtle-Duck/DALI_video_test), which gets minor different results. 

I test CPU version in different FFmpeg versions which have same results, so i also guess decoding implemented in GPU leads this difference.

**Comment by JanuszL** [comment]
Thanks, we will look into it.

**Comment by Turtle-Duck** [comment]
Thanks you very much!

**Comment by JanuszL** [comment]
I have managed to reproduce your values using DALI (they seem to be the same as for pure NVVL) using following script:
```
# python 3.4

from __future__ import print_function
from __future__ import division
import os
import numpy as np
from PIL import Image
from nvidia.dali.pipeline import Pipeline
import nvidia.dali.ops as ops
import nvidia.dali.types as types
import imageio
from subprocess import call

path = 'video_data/7275702640_b.mp4'
vid = imageio.get_reader(path, 'ffmpeg')
call(['video_data/ffmpeg/ffmpeg', '-y', '-hwaccel', 'nvdec', '-i', path, 'video_data/%03d.bmp'])

class VideoPipe(Pipeline):
    def __init__(self, batch_size, num_threads, device_id, data, shuffle):
        super(VideoPipe, self).__init__(batch_size, num_threads, device_id, seed=16)
        self.input = ops.VideoReader(device="gpu", filenames=data, sequence_length=1,
                                     shard_id=0, num_shards=1,
                                     random_shuffle=shuffle, initial_fill=1)

    def define_graph(self):
        output = self.input(name="Reader")
        return output

pipe = VideoPipe(batch_size=1, num_threads=4, device_id=0, data=path, shuffle=False)
pipe.build()

for index in range(5):
    sequences_out = pipe.run()[0].as_cpu().as_array()
    image_dali = Image.fromarray(sequences_out[0][0], 'RGB')
    image_dali.save('dali{}.png'.format(index))
    image_dali = np.asarray(image_dali)

    frame = vid.get_data(index)[:, :, [2, 1, 0]]
    b, g, r = Image.fromarray(frame, 'RGB').split()
    image_ffmpeg = Image.merge("RGB", (r, g, b))
    image_ffmpeg.save('ffmpeg{}.png'.format(index))
    image_ffmpeg = np.asarray(image_ffmpeg)

    reference = imageio.imread('video_data/00{}.bmp'.format(index+1))
    reference = np.asarray(reference)

    diff_image = (image_ffmpeg - image_dali) * np.uint8(image_ffmpeg<image_dali) * 254 + 1
    Image.fromarray(diff_image, 'RGB').save('diff{}.bmp'.format(index))
    print("DALI")
    print(image_dali)
    print("FFMPEG")
    print(image_ffmpeg)
    print("REFERENCE")
    print(reference)
    print("***"*10)
```
In `video_data/ffmpeg/ffmpeg` path I have ffmpeg compiled with CUDA enabled (please check [this guide](https://developer.nvidia.com/ffmpeg)).

I see some discrepancies, mostly off by 1 (between DALI based on NVVL, imageio FFmpeg and FFmpeg with NVDEC acceleration enabled). I will keep digging.

**Comment by Turtle-Duck** [comment]
Thanks for your script！I will also dig this discrepancies, thank you.

**Comment by JanuszL** [comment]
Hi,
https://github.com/NVIDIA/DALI/pull/1054 partially addresses that problem - it fixes wrong sampling of chroma channel during conversion from raw decoded frames to the output memory. If the interpolation strategy is set to the nearest (manual change in code done during testing)  it returns bit-exact raw frames comparing with the latest FFmpeg compiled from sources. We will keep interpolation as linear due to a subjective better quality of the output images.
Also, there are some discrepancies connected with the fact of how colors are converted between YUV and RGB but this is also the matter of the implementation.
To sum up I wouldn't worry about remaining discrepancies after https://github.com/NVIDIA/DALI/pull/1054  is merged.

**Comment by Turtle-Duck** [comment]
Thanks! thats very helpful.

**Comment by JanuszL** [comment]
It should be included in 0.12.0
